tango/integrations/beaker/workspace.py

```diff
@@ -0,0 +1,278 @@
+        elif parsed_url.netloc:
+            # e.g. "beaker://my-workspace"
+            workspace = parsed_url.netloc
+        else:
+            raise ValueError(f"Bad URL for Beaker workspace '{parsed_url}'")
+        return cls(workspace)
+
+    @property
+    def step_cache(self) -> StepCache:
+        return self.cache
+
+    def step_dir(self, step_or_unique_id: Union[Step, str]) -> Path:
+        unique_id = (
+            step_or_unique_id if isinstance(step_or_unique_id, str) else step_or_unique_id.unique_id
+        )
+        path = self.steps_dir / unique_id
+        path.mkdir(parents=True, exist_ok=True)
+        return path
+
+    def work_dir(self, step: Step) -> Path:
+        path = self.step_dir(step) / "work"
+        path.mkdir(parents=True, exist_ok=True)
+        return path
+
+    def step_info(self, step_or_unique_id: Union[Step, str]) -> StepInfo:
+        try:
+            dataset = self.beaker.dataset.get(step_dataset_name(step_or_unique_id))
+            file_info = self.beaker.dataset.file_info(dataset, Constants.STEP_INFO_FNAME)
+            step_info: StepInfo
+            if file_info.digest in self._step_info_cache:
+                step_info = self._step_info_cache.pop(file_info.digest)
+            else:
+                step_info_bytes = b"".join(
+                    self.beaker.dataset.stream_file(dataset, file_info, quiet=True)
+                )
+                step_info = StepInfo.from_json_dict(json.loads(step_info_bytes))
+            self._step_info_cache[file_info.digest] = step_info
+            while len(self._step_info_cache) > self.STEP_INFO_CACHE_SIZE:
+                self._step_info_cache.popitem(last=False)
+            return step_info
+        except (DatasetNotFound, FileNotFoundError):
+            raise KeyError(step_or_unique_id)
+
+    def step_starting(self, step: Step) -> None:
+        # We don't do anything with uncacheable steps.
+        if not step.cache_results:
+            return
+
+        lock_path = self.step_dir(step) / "lock"
+        lock = FileLock(lock_path, read_only_ok=True)
+        lock.acquire_with_updates(desc=f"acquiring lock for '{step.name}'")
+        self.locks[step] = lock
+        step_info = self.step_info(step)
+        if step_info.state not in {StepState.INCOMPLETE, StepState.FAILED, StepState.UNCACHEABLE}:
+            raise RuntimeError(
+                f"Step '{step.name}' is trying to start, but it is already {step_info.state}. "
+                f"If you are certain the step is not running somewhere else, delete the lock step info "
+                f"Beaker dataset at {self.beaker.dataset.url(step_dataset_name(step))}"
+            )
+
+        # Update StepInfo to mark as running.
+        try:
+            step_info.start_time = utc_now_datetime()
+            step_info.end_time = None
+            step_info.error = None
+            step_info.result_location = None
+            self._update_step_info(step_info)
```

Comment:
There is a race condition here. Do we have that race condition in the other workspaces as well?

Reply:
You're right. I have an idea for a remote lock on Beaker. `LocalWorkspace` doesn't have this issue of course, but `WandbWorkspace` might.

Reply:
There's another issue here: Workspaces in generally don't handle these race conditions well. They'll also result in a `RuntimeError` or some other (potentially silent) failure. This is my proposed fix for that: e1bf551

Reply:
Remote lock: 1bc7cc3